File: app/src/lib/sprachnotiz.ts
/**
 * Eine Sprachnotiz aufnehmen und auf die Boxen legen.
 *
 * «Levin, in fünf Minuten gehen wir» hat man schneller gesagt als
 * getippt – und aus der Box klingt es nach jemandem statt nach der
 * Vorlesestimme. Der Weg dahinter ist derselbe wie bei der Durchsage
 * (hub/core/sprachnotiz.py): aufnehmen, hochladen, abspielen.
 *
 * **Zwei Wege zum selben Ergebnis.** Im Browser und auf dem Wandpanel
 * nimmt der `MediaRecorder` auf, auf dem Telefon und dem iPad
 * `expo-audio` (lib/aufnahme-nativ.ts). Herauskommt beide Male ein
 * Blob, den der Hub auf die Boxen legt.
 *
 * Lange gab es nur den Browser-Weg, und das aus einem guten Grund: Ein
 * natives Aufnahmemodul heisst eine neue `runtimeVersion`, und die
 * kappt den OTA-Kanal für *alle* Telefone, bis jemand einen
 * TestFlight-Build installiert (CLAUDE.md, «Ausliefern»). Das war ein
 * hoher Preis für eine Kleinigkeit.
 *
 * Bezahlt wurde er trotzdem – aber nicht dafür: `expo-quick-actions`
 * kam als natives Modul dazu und machte einen neuen Build ohnehin
 * fällig. In derselben Runde kostet das Mikrofon nichts mehr extra.
 *
 * Wo es *doch* keines gibt (ein alter Browser ohne `MediaRecorder`),
 * zeigt die App den Knopf gar nicht erst, statt ihn auszugrauen: Ein
 * Knopf, der nie geht, ist ein Versprechen, das die App nicht hält.
 */
import { Platform } from 'react-native';


/** So lange darf eine Notiz höchstens laufen - danach stoppt sie
 *  selbst. Ein Zuruf, kein Vortrag; der Hub weist längeres ohnehin ab. */
export const HOECHSTENS_MS = 60_000;

/** Darunter war es ein Verrutschen am Knopf und keine Aufnahme. */
export const MINDESTENS_MS = 700;

/** Was der Browser aufnehmen soll, in der Reihenfolge der Vorliebe.
 *  Opus in WebM spielen Google-Cast-Boxen; das ist der Normalfall hier. */
const FORMATE = ['audio/webm;codecs=opus', 'audio/webm', 'audio/mp4', 'audio/ogg'];

/**
 * Gibt es hier überhaupt ein Mikrofon, an das die App kommt?
 *
 * Nativ ja: Das Modul ist Teil der App-Hülle, und ob der Benutzer das
 * Mikrofon freigibt, entscheidet er beim ersten Druck - nicht hier.
 * Danach zu fragen, bloss um den Knopf zu zeigen, hiesse das Fenster
 * «HomePilot möchte auf das Mikrofon zugreifen» aufzumachen, bevor
 * jemand etwas sagen wollte.
 */
export function kannAufnehmen(): boolean {
  if (Platform.OS !== 'web') return true;
  return (
    typeof navigator !== 'undefined' &&
    !!navigator.mediaDevices?.getUserMedia &&
    typeof MediaRecorder !== 'undefined'
  );
}

/** Das erste Format, das dieser Browser kann - oder nichts (rein, testbar). */
export function bestesFormat(
  moeglich: (typ: string) => boolean,
  formate: string[] = FORMATE
): string | undefined {
  return formate.find((typ) => {
    try {
      return moeglich(typ);
    } catch {
      return false;
    }
  });
}

/**
 * «0:07» – die laufende Aufnahmedauer (rein, testbar).
 *
 * Mit Sekunden und nicht nur als Punkt: Wer aufnimmt, hat keinen
 * anderen Anhalt dafür, ob das Mikrofon wirklich läuft.
 */
export function dauerText(ms: number): string {
  const gesamt = Math.max(0, Math.floor(ms / 1000));
  return `${Math.floor(gesamt / 60)}:${String(gesamt % 60).padStart(2, '0')}`;
}

/**
 * Warum die Aufnahme nicht losging, auf Deutsch (rein, testbar).
 *
 * Die Namen kommen vom Browser und sagen einem Menschen nichts. Der
 * häufigste Fall ist der erste: Die Erlaubnis wurde abgelehnt oder das
 * Fenster mit der Frage weggeklickt.
 */
export function aufnahmeFehler(err: unknown): string {
  const name = (err as { name?: string })?.name ?? '';
  if (name === 'NotAllowedError' || name === 'SecurityError') {
    // Zwei Welten, zwei Orte: Im Browser fragt die Adressleiste, auf dem
    // Telefon steht es in den Einstellungen. «Im Browser erlauben» half
    // dort niemandem.
    return Platform.OS === 'web'
      ? 'Das Mikrofon ist nicht freigegeben - im Browser erlauben und noch einmal versuchen.'
      : 'Das Mikrofon ist nicht freigegeben - in den Einstellungen des Telefons unter HomePilot erlauben.';
  }
  if (name === 'NotFoundError' || name === 'OverconstrainedError') {
    return 'Kein Mikrofon gefunden.';
  }
  if (name === 'NotReadableError') {
    return 'Das Mikrofon ist gerade von einem anderen Programm belegt.';
  }
  return 'Die Aufnahme ging nicht los.';
}

export interface Aufnahme {
  /** Beenden und die Aufnahme liefern. `null`, wenn nichts ankam. */
  stopp: () => Promise<Blob | null>;
  /** Wegwerfen - das Mikrofon wird trotzdem freigegeben. */
  abbrechen: () => void;
}

/**
 * Aufnahme starten. Wirft mit lesbarem Grund (siehe aufnahmeFehler).
 *
 * Die Spur wird am Ende ausdrücklich beendet: Sonst bleibt im Browser
 * der rote Aufnahmepunkt im Tab stehen, und auf dem Wandpanel sähe es
 * aus, als höre die Wohnung dauerhaft zu.
 *
 * Auf dem Telefon übernimmt `aufnahme-nativ.ts`. Nachgeladen und nicht
 * oben importiert: Sonst zöge der Web-Bau das native Modul mit.
 */
export async function starteAufnahme(): Promise<Aufnahme> {
  if (Platform.OS !== 'web') {
    const { starteNativ } = await import('./aufnahme-nativ');
    return starteNativ(HOECHSTENS_MS);
  }
  const spur = await navigator.mediaDevices.getUserMedia({ audio: true });
  const typ = bestesFormat((t) => MediaRecorder.isTypeSupported(t));
  const recorder = new MediaRecorder(spur, typ ? { mimeType: typ } : undefined);
  const stuecke: Blob[] = [];
  recorder.ondataavailable = (ereignis) => {
    if (ereignis.data && ereignis.data.size > 0) stuecke.push(ereignis.data);
  };
  recorder.start();

  const aufraeumen = () => spur.getTracks().forEach((track) => track.stop());

  // Selbstabschaltung: Wer den Knopf vergisst, soll nicht nach zehn
  // Minuten eine Aufnahme haben, die der Hub ohnehin abweist.
  const frist = setTimeout(() => {
    if (recorder.state === 'recording') recorder.stop();
  }, HOECHSTENS_MS);

  return {
    stopp: () =>
      new Promise<Blob | null>((fertig) => {
        clearTimeout(frist);
        const liefern = () => {
          aufraeumen();
          fertig(stuecke.length > 0 ? new Blob(stuecke, { type: typ ?? 'audio/webm' }) : null);
        };
        if (recorder.state === 'inactive') {
          liefern();
          return;
        }
        recorder.onstop = liefern;
        recorder.stop();
      }),
    abbrechen: () => {
      clearTimeout(frist);
      if (recorder.state !== 'inactive') recorder.stop();
      aufraeumen();
    },
  };
}

File: hub/homepilot/core/sprachnotiz.py
"""Sprachnotiz: einen selbst gesprochenen Satz auf die Boxen legen.

Eine Durchsage schreibt man («Essen ist fertig»), und der Hub spricht
sie vor. Das ist gut für die immergleichen Sätze und schlecht für alles
andere: Wer «Levin, in fünf Minuten gehen wir» tippt, hat länger
getippt, als er gebraucht hätte, es zu sagen – und die Stimme aus der
Box klingt nach niemandem.

Deshalb hier derselbe Weg, nur mit dem Mikrofon davor: aufnehmen,
hochladen, auf die Boxen legen (say.play_audio). Der Hub wandelt nichts
um; er reicht durch, was ankommt, und nennt der Box den richtigen Typ.
Umwandeln hiesse ffmpeg auf dem Hub, und das ist ein grosses Stück
Software für einen Satz.

**Was die Boxen davon spielen.** Google-Cast-Geräte spielen WebM/Opus –
also genau das, was ein Browser aufnimmt. Ältere Boxen wollen MP3 oder
WAV. Was nicht spielt, ist an der Box still und steht als Fehler in der
Antwort; der Hub kann das nicht vorher wissen.

**Wo aufgenommen wird.** Überall: im Browser und auf dem Wandpanel mit
dem ``MediaRecorder``, auf Telefon und iPad mit ``expo-audio``. Lange
gab es nur den ersten Weg, weil ein natives Aufnahmemodul eine neue
``runtimeVersion`` bedeutet – und die kappt den OTA-Kanal für alle
Telefone, bis jemand einen TestFlight-Build installiert (CLAUDE.md,
«Ausliefern»). Bezahlt wurde dieser Preis dann für etwas anderes
(``expo-quick-actions``); in derselben Runde kostete das Mikrofon nichts
mehr extra.

Vom Telefon kommt AAC in einem MP4-Rumpf statt WebM/Opus. ``medientyp``
erkennt beides; Google-Cast-Boxen spielen beides.
"""

from __future__ import annotations

from .errors import HomePilotError

#: Kürzer ist keine Aufnahme, sondern ein Verrutschen am Knopf.
MINDESTENS = 800

#: Rund eine Minute Opus. Eine Sprachnotiz ist ein Zuruf, kein Vortrag -
#: und der Ton liegt bis zum Abspielen im Arbeitsspeicher des Hubs.
HOECHSTENS = 2_000_000

#: Anfangsbytes → Medientyp. Ein Dateiname existiert hier nicht; die
#: Aufnahme kommt als roher Rumpf an.
KOEPFE: tuple[tuple[bytes, str], ...] = (
    (b"RIFF", "audio/wav"),
    (b"OggS", "audio/ogg"),
    (b"\x1a\x45\xdf\xa3", "audio/webm"),
    (b"ID3", "audio/mpeg"),
    (b"\xff\xfb", "audio/mpeg"),
    (b"\xff\xf3", "audio/mpeg"),
)


def medientyp(audio: bytes) -> str:
    """Was für ein Ton das ist (rein, testbar).

    Die Box bekommt den Typ als Content-Type und richtet sich danach.
    Vorher stand dort «audio/mpeg» für alles, was nicht mit RIFF anfing -
    bei einer WebM-Aufnahme ist das schlicht gelogen, und eine Box, die
    einer falschen Angabe glaubt, spielt Rauschen oder gar nichts.
    """
    for kopf, typ in KOEPFE:
        if audio.startswith(kopf):
            return typ
    # MP4/M4A trägt seine Kennung erst an Position vier.
    if len(audio) > 11 and audio[4:8] == b"ftyp":
        return "audio/mp4"
    # Unbekannt: MP3 ist die wahrscheinlichste Annahme und die, mit der
    # die meisten Boxen etwas anfangen.
    return "audio/mpeg"


def pruefen(audio: bytes) -> None:
    """Grösse prüfen, sonst nichts. Wirft HomePilotError mit Klartext.

    Bewusst keine Formatprüfung: Welche Formate ein Browser aufnimmt,
    ändert sich mit jeder Version, und eine Liste, die der Hub pflegen
    müsste, wäre in einem Jahr falsch. Was die Box nicht spielt, meldet
    sie - das ist die ehrlichere Grenze.
    """
    if len(audio) < MINDESTENS:
        raise HomePilotError(
            "Die Aufnahme ist zu kurz - hat der Knopf gewackelt?"
        )
    if len(audio) > HOECHSTENS:
        raise HomePilotError(
            "Die Aufnahme ist zu lang - eine Sprachnotiz ist ein Zuruf "
            "(höchstens etwa eine Minute)."
        )
